### PR TITLE
Properly clone elements in Spl lists/stacks

### PR DESCRIPTION
**Pull request by michaelmoussa**
I happened to run into this when attempting to clone an object that has an `SplStack` somewhere in its object association graph. The default behavior of `DeepCopy` creates a cloned `SplStack` such that `$stack == $clonedStack` and `$stack !== $clonedStack`, which is fine; however, the items _within_ the stack are not deep-copied... that is, `$stack->pop() === $clonedStack->pop()`.

This PR solves that issue (for `SplDoublyLinkedList`, which solves it for `SplStack` as well) by taking advantage of filters and matches. I'm not sure if adding the typefilter to the constructor is an acceptable way to do this (I generally avoid `new` in constructors myself), but I can't think of a reason why you _wouldn't_ want to properly deep-copy stacks and linked lists, given the name of this library.

I suspect data structures like SPL heaps, queues, etc... may be affected in the same way, but my immediate need is for stacks and figured I'd share what I've done to solve that.

**tl;dr:** uncomment the new line 51 of `src/DeepCopy/DeepCopy.php` and run PHPUnit. :)

### CODE REVIEW

**Comment by mnapoli** [comment]
Hi, I'm not sure about adding support for native PHP classes one at a time, and I'm afraid it might break stuff for existing users who might have already filters for them.

Maybe it's worth adding full support to all native PHP classes and doing that in a new major version? That would be interesting enough for a new major version, plus it would be also the opportunity to clean up a few things/hacks that have piled up since 1.0 (@theofidry if you have an opinion on that)

**Comment by michaelmoussa** [comment]
@mnapoli I think it'd be worth it for a major version as well, but the trouble would be identifying which native PHP classes are not being "properly" cloned, and where to draw the line. There are hundreds of native classes (assuming you don't take modules into consideration), most of which probably work just fine.

In the case of these Spl datastructures, I only happened to notice because one of the items in the structure was stateful and not reusable once invoked. I suspect that might be the case for most affected classes, otherwise someone would have opened an issue by now (I would hope!).

In the meantime, perhaps making the filter available but _not_ adding it automatically in the constructor, plus adding a note to the docs, would be a happy medium for making a fix available immediately while not breaking BC for anyone?

**Comment by mnapoli** [comment]
> but the trouble would be identifying which native PHP classes are not being "properly" cloned, and where to draw the line.

> […]

> Worthwhile effort in the long run, but I think there'd need to be a clear decision on the extent and approach.

Agreed.

> In the meantime, perhaps making the filter available but not adding it automatically in the constructor, plus adding a note to the docs, would be a happy medium for making a fix available immediately while not breaking BC for anyone?

Yeah very good idea! I'm :+1: with that

**Comment by michaelmoussa** [comment]
> Yeah very good idea! I'm 👍 with that

Great! I've updated the PR to fix the PHP 5.4 build issue, so this should be good to go whenever you're ready.

**Comment by mnapoli** [comment]
Thank you, sorry for the delay!

**Comment by mnapoli** [comment]
I've tagged [1.5.5](https://github.com/myclabs/DeepCopy/releases/tag/1.5.5)

**Comment by michaelmoussa** [comment]
@mnapoli no problem - thanks for getting to it! :)
